Add support for automatic port-forwarding in Hubble CLI

devodev opened the pull request:
Implement automatic (opt-in) port-forwarding capability in Hubble-cli.

As proposed in https://github.com/cilium/cilium/pull/35353, this uses native port-forwarding by introducing a new high-level PortForwarder type to add automatic port-forwarding to all commands that require a GRPC connection to hubble-relay.

Fixes: #34935

```release-note
Add support for automatic port-forwarding in Hubble CLI
Replace kubectl-based port-forwarding with native implementation in Cilium CLI
```

---

This also updates hubble port-forwarding commands in cilium-cli to replace the kubectl-based implementation, but we can also do that in a different PR if desired. Added bonus is we don't need to sleep in a goroutine before opening the browser to mitigate the kubectl spawn process delay, nice.

devodev committed:
cilium: vendor k8s.io/kubectl/pkg/util and k8s.io/kubectl/pkg/util/podutils

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev committed:
cilium-cli: use native port-forwarding in hubble commands

Replace kubectl-based port-forwarding with a native
implementation inspired by the kubectl port-forward
cmd introduced by the PortForwarder helper type.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev commented on hubble/cmd/common/config/flags.go:
```diff
@@ -28,6 +28,11 @@ const (
+	KeyPortForward       = "port-forward"         // uint16
+	KeyK8sContextName    = "k8s-context"          // string
+	KeyK8sNamespace      = "k8s-namespace"        // string
+	KeyK8sKubeconfig     = "k8s-kubeconfig"       // string
```
If you have a suggestion for alternative names, don't hesitate.

I also though of grouping port-forward flags together similar to `--tls` like:
```
      --port-forward                  Automatically forward the relay port to the local machine. Analoguous to running: 'cilium hubble port-forward'.
      --port-forward-port uint16      Local port to forward to. 0 will select a random port. This option is only considered when --auto-port-forward is set. (default 4245)
```

chancez replied:
I like `--port-forward-port` more than `--port-forward` for configuring the port. Otherwise I think the rest is fine. 

If I was to nit-pick`--k8s-kubeconfig` could probably be just `--kubeconfig` since we don't have to worry about the potential conflict if we add config/contexts for hubble CLI. I could also argue `--kube-context`/`--kube-namespace` instead of `--k8s-context`/`--k8s-namespace` which is the naming scheme argocd CLI uses, but I don't have strong feelings about that. Ill defer to you on those.

devodev replied:
I like your suggestions, will update, thanks

chancez commented on pkg/common/k8s/dialer.go:
```diff
@@ -101,3 +109,46 @@ func (c *Client) PortForward(ctx context.Context, p PortForwardParameters) (*Por
+//
+// To stop the port-forwarding, use the context by cancelling it.
+func (c *Client) PortForwardService(ctx context.Context, namespace, name string, localPort, svcPort int32) (*PortForwardServiceResult, error) {
+	relaySvc, err := c.GetService(ctx, namespace, name, metav1.GetOptions{})
```
Nit: `relaySvc` isn't accurate anymore. Could probably just be `svc`

chancez reviewed (approved):
LGTM overall. Mostly nitpicks so nothing blocking and the functional changes are good. Thanks for keeping things broken into multiple commits.

derailed reviewed (approved):
@devodev Nice work!

gandro commented on pkg/cli/k8s/helpers.go:
```diff
@@ -101,3 +105,29 @@ func (c *Client) Describe(obj runtime.Object) (gvk schema.GroupVersionKind, reso
+
+	pods := make([]*corev1.Pod, 0, len(podList.Items))
+	for _, pod := range podList.Items {
+		pod := pod
```
This is no longer needed in Go 1.23

devodev replied:
Haha you're right! habits persist, let me change that

devodev replied:
Fixed [here](https://github.com/cilium/cilium/pull/35483/commits/ec079ad78661d41fa984a80ab2b17396acce9418).

gandro reviewed (approved):
Excellent work! One very minor nit

devodev committed:
hubble: add support for automatic port-forwarding

Add the capability to connect to hubble relay by creating
a port-forward to the appropriate pod similar to manually
spawning a new process with 'cilium hubble port-forward'.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev committed:
cilium-cli: extract portforward logic into new PortForwarder type

Add a new type that provides helper methods for kubernetes
port-forwarding which is not linked to the cilium-cli k8s
client to allow re-use.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev committed:
k8s: add PortForwardService method to PortForwarder

Add a new helper method that finds the best candidate
pod for port-forwrding using a service reference.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>
